src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1250,34 +1317,103 @@ public void OnStartLine(HttpMethod method, HttpVersion version, Span<byte> targe
+        }
+
+
+        private void OnAuthorityFormTarget(HttpMethod method, Span<byte> target, Span<byte> line)
+        {
+            // TODO Validate that target is a correct host[:port] string.
+            // Reject as 400 if not. This is just a quick scan for invalid characters
+            // but doesn't check that the target fully matches the URI spec.
+            for (var i = 0; i < target.Length; i++)
+            {
+                var ch = (char)target[i];
+                if (!char.IsLetterOrDigit(ch) && ch != ':' && ch != '.')
+                {
+                    if (Log.IsEnabled(LogLevel.Information))
+                    {
+                        RejectRequestLine(line);
+                    }
+
+                    throw BadHttpRequestException.GetException(RequestRejectionReason.InvalidRequestLine);
+                }
+            }
+
+            // The authority-form of request-target is only used for CONNECT
+            // requests (https://tools.ietf.org/html/rfc7231#section-4.3.6).
+
+            if (method != HttpMethod.Connect)
             {
-                Method = HttpUtilities.MethodToString(method) ?? string.Empty;
+                RejectRequest(RequestRejectionReason.ConnectMethodRequired);
             }
-            else
+
+            // When making a CONNECT request to establish a tunnel through one or
+            // more proxies, a client MUST send only the target URI's authority
+            // component(excluding any userinfo and its "@" delimiter) as the
+            // request - target.For example,
+            //
+            //  CONNECT www.example.com:80 HTTP/1.1
+            //
+            // Allowed characters in the 'host + port' section of authority.
+            // See https://tools.ietf.org/html/rfc3986#section-3.2
+
+            RawTarget = target.GetAsciiStringNonNullCharacters();
+            Path = string.Empty;
+            PathBase = string.Empty;
+            QueryString = string.Empty;
+        }
+
+        private void OnAsteriskFormTarget(HttpMethod method)
+        {
+            // The asterisk-form of request-target is only used for a server-wide
+            // OPTIONS request (https://tools.ietf.org/html/rfc7231#section-4.3.7).
+            if (method != HttpMethod.Options)
             {
-                Method = customMethod.GetAsciiStringNonNullCharacters();
+                RejectRequest(RequestRejectionReason.OptionsMethodRequired);
             }
 
+            RawTarget = Asterisk;
+            Path = string.Empty;
+            PathBase = string.Empty;
+            QueryString = string.Empty;
+        }
+
+        private void OnAbsoluteFormTarget(Span<byte> target, Span<byte> query, Span<byte> line)
+        {
+            // absolute-form
+            // https://tools.ietf.org/html/rfc7230#section-5.3.2
+
+            RawTarget = target.GetAsciiStringNonNullCharacters();
+
+            if (!Uri.TryCreate(RawTarget, UriKind.Absolute, out var uri))
```

Comment:
This code is about correctness, not perf. For context, we talked about this in the [first iteration](https://github.com/aspnet/KestrelHttpServer/pull/1246) of absolute-URI support. We're aware this is slow, but it's not on the hot path. In fact, it's not really on any commonly used code path. The HTTP spec requires we handle requests like `GET http://localhost/path HTTP/1.1`, but in practice, clients never do. Instead they use origin-form, `GET /path HTTP/1.1`. We debated whether to bother at all, but since our goal is edge-server capabilities, we will comply with the HTTP spec, including its edge-cases. Fast URI parsing is not a high priority.